src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -134,25 +134,21 @@ public UnfilteredPartitionIterator search(ReadExecutionController executionContr
             queryContext.addShadowedKeysLoopCount(1L);
             long lastShadowedKeysCount = queryContext.getShadowedPrimaryKeys().size();
             ResultRetriever result = queryIndexes.get();
-            queryContext.resetRowsMatched();
             UnfilteredPartitionIterator topK = (UnfilteredPartitionIterator) new VectorTopKProcessor(command).filter(result);
             long currentShadowedKeysCount = queryContext.getShadowedPrimaryKeys().size();
             long newShadowedKeysCount = currentShadowedKeysCount - lastShadowedKeysCount;
             logger.debug("Shadow loop iteration {}: rows matched: {}, keys shadowed: {}",
                          queryContext.shadowedKeysLoopCount(),
-                         queryContext.rowsMatched(),
+                         Math.max(0, queryContext.softLimit() - currentShadowedKeysCount),
                          currentShadowedKeysCount);
-            // Stop if no new shadowed keys found or if we already got enough rows
-            if (newShadowedKeysCount == 0 || exactLimit <= queryContext.rowsMatched())
+            // Stop if no new shadowed keys found or if we already got enough rows (In this case, we're basically
+            // checking that if all the shadowed keys came from one index, we still got enough non-shadowed rows from
+            // that index to still guarantee we have a valid result.)
+            if (newShadowedKeysCount == 0 || exactLimit <= queryContext.softLimit() - currentShadowedKeysCount)
```

Comment:
Actually, this does re-introduce the perf regression. I need to look closer at these commits to understand what the soft limit improvements were in #927.